### PR TITLE
stage-1: Log to secondary consoles

### PR DESCRIPTION
**Pull request by samueldr**
### Motivation for this change

The output from `init` will only be present on the last valid `console=` parameter from the kernel command-line.

This is extremely inconvenient, when not harmful, for generic images. On many ARM platforms the serial console names differ, so there is no absolute truthful console name to use here. Additionally, this prevents logging to be shown on both the connected display, and the serial output at once.

With `bootlogd`, the output is redirected to all `console=` parameters. It also replaces the ad-hoc logging to /run/log and the logging to /dev/kmsg.

This way we should be able to get rid of one of the main annoyance with ARM platforms early boot debugging.

This has been used for a good while in Mobile NixOS.

Ideally we'd somehow do a system test that confirms it logs to multiple serial consoles, but I really don't know how we'd pull that off.

### TODO

 - [ ] Investigate input redirection
 - [ ] Decide how we handle killing bootlogd
 - [ ] Verify operation on an ARM machine

#### Input redirection

While opening this PR, I realized that this probably doesn't cover input during stage-1. Only the last `console=` would be usable as an input in that case. This is important for `boot.debug1*` use cases, and also useful if only to answer `r` to reboot the machine.

We probably can, could, and should merge this even without this being handled, as it would already be a net improvement compared to the current situation.

#### Killing bootlogd

Right now I haven't handled it at all. It gets culled at `Kill any remaining processes`, which means it will be missing all messages up to systemd starting in stage-2.

Should we:

 - Cheat and mark it as a storage daemon by prefixing its argv0 with a `@`?
 - Special case it in the loop?

Then, we can add a late `boot.postBootCommands` which includes `${pkgs.procps}/bin/pkill -x bootlogd` to kill it. Ideally order it to happen last before we `exec` into systemd.

### What even is `bootlogd`??

 - https://github.com/mobile-nixos/bootlogd

It's a minified part of `sysvinit`. The main core hasn't been touched by me. Really the main thing I did was delete the `sysvinit` around it. Then I added features that helps either Mobile NixOS or NixOS in some ways.

### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

* * *

This closes #42255.

### CODE REVIEW

**Comment by grahamc** [comment]
> While opening this PR, I realized that this probably doesn't cover input during stage-1. Only the last console= would be usable as an input in that case. This is important for boot.debug1* use cases, and also useful if only to answer r to reboot the machine.

It might be useful to print a message like "This is <currentconsoledevice>, input is only accepted on <consoledevice>.".

I'm anxious about merging this in before 21.05 branchoff so soon to cutoff, but would be +1 on the next release.

**Comment by samueldr** [comment]
> It might be useful to print a message like "This is , input is only accepted on .".

`bootlogd` doesn't know about any input at this time. So it wouldn't really be able to. Adding support for it is probably not that much less work than adding input multiplexing. Additionally, it wouldn't really be able to say it at a useful time. It would likely state it early during boot. So a few dozen stage-1 lines, probably combined with dozens of kernel log, it wouldn't be visible when input is needed.

The stage-1 sh script *could* loop on `console=` parameters to state so. It would know when input is handed-off to the user.

File: nixos/modules/system/boot/stage-1-init.sh [hunk 1]
```diff
@@ -15,6 +15,21 @@ export LD_LIBRARY_PATH=@extraUtils@/lib
 export PATH=@extraUtils@/bin
 ln -s @extraUtils@/bin /bin
 
+
+# `-f true` to force bootlogd to fork itself. bootlogd will only fork once it's ready to consume logs.  Otherwise it could skip the first few lines.
+# `-p` ensures required mount points are mounted.
+# `-l` logs to /dev/kmsg too
```
Comment (samueldr) [inline nixos/modules/system/boot/stage-1-init.sh:21]:
```suggestion
# `-k` logs to /dev/kmsg too
```

**Comment by samueldr** [comment]
Closing in favour of systemd's stage-1 infra, which I assume intrinsically supports console multiplexing just like its stage-2 does.

I won't get back to dealing with the complexity of this.

**Pull request closed without merging**